File: coding/src/com/algos/core/li10_tree/bt/revised/Bt_L987_VerticalOrderTraversalOfABinaryTree.java
package  com.algos.core.li10_tree.bt.revised;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.TreeMap;

import  com.algos.core.li30_model.TreeNode;

public class Bt_L987_VerticalOrderTraversalOfABinaryTree {
    public static void main(String[] args) {
        Bt_L987_VerticalOrderTraversalOfABinaryTree obj = new Bt_L987_VerticalOrderTraversalOfABinaryTree();

        TreeNode root = new TreeNode(3);
		root.left = new TreeNode(9);
        root.right = new TreeNode(20);
		root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println(obj.verticalTraversal(root));
    }

    // Problem: https://leetcode.com/problems/vertical-order-traversal-of-a-binary-tree/
    // Idea: Map each node to 2d plane row, col coordinates. Store row, cols in queue while doing bfs.
    // Group all the cols and respect rows in TreeMaps.
    // Time: O(n)
    // space: O(n)
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        TreeMap<Integer, TreeMap<Integer, PriorityQueue<Integer>>> map = new TreeMap<>();
        Queue<Tuple> q = new LinkedList<>();
        q.add(new Tuple(root, 0, 0));
        while(!q.isEmpty()){
            Tuple curr = q.remove();
            TreeNode node = curr.node;
            int row = curr.row, col = curr.col;
            
            if(!map.containsKey(col)) 
                map.put(col, new TreeMap<>());
            if(!map.get(col).containsKey(row)) 
                map.get(col).put(row, new PriorityQueue<Integer>());
            map.get(col).get(row).add(node.val);

            if(node.left != null) 
                q.add(new Tuple(node.left, row+1, col-1));
            if(node.right != null) 
                q.add(new Tuple(node.right, row+1, col+1));
        }

        List<List<Integer>> ans = new ArrayList<>();
        for(TreeMap<Integer, PriorityQueue<Integer>> cols: map.values()){
            ans.add(new ArrayList<>());
            for(PriorityQueue<Integer> nodes: cols.values()){
                while(!nodes.isEmpty()){
                    ans.get(ans.size()-1).add(nodes.remove());
                }
            }
        }

        return ans;
    }
}

class Tuple{
    public TreeNode node;
    public int row;
    public int col;
    public Tuple(TreeNode node, int row, int col) {
        this.node = node;
        this.row = row;
        this.col = col;
    }
    
}

File: dsa/concepts/10_trees/bt/python/bt_l987__vertical_order_traversal_of_a_binary_tree.py
# Binary Tree + BFS + Sorting
# Problem: https://leetcode.com/problems/vertical-order-traversal-of-a-binary-tree/
# Idea:
# 1. Map each node to 2D coordinates (row, col) using BFS.
# 2. Use nested dictionaries: col -> row -> min-heap of node values.
# 3. After BFS, extract values in order: sorted columns, sorted rows, sorted node values.
# Time: O(n log n) (due to sorting of rows and min-heaps)
# Space: O(n)

from collections import deque, defaultdict
import heapq

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class VerticalTuple:
    def __init__(self, node, row, col):
        self.node = node
        self.row = row
        self.col = col

def vertical_traversal(root):
    if not root:
        return []

    # col -> row -> min-heap of node values
    node_map = defaultdict(lambda: defaultdict(list))
    q = deque([VerticalTuple(root, 0, 0)])

    while q:
        curr = q.popleft()
        node, row, col = curr.node, curr.row, curr.col
        heapq.heappush(node_map[col][row], node.val)

        if node.left:
            q.append(VerticalTuple(node.left, row + 1, col - 1))
        if node.right:
            q.append(VerticalTuple(node.right, row + 1, col + 1))

    ans = []
    for col in sorted(node_map.keys()):
        col_nodes = []
        for row in sorted(node_map[col].keys()):
            while node_map[col][row]:
                col_nodes.append(heapq.heappop(node_map[col][row]))
        ans.append(col_nodes)

    return ans

# -------- Test --------
root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

print(vertical_traversal(root))  # Expected: [[9], [3, 15], [20], [7]]
